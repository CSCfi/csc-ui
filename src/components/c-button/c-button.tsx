import { Component, Prop, h } from '@stencil/core';
import { mdiPlus, mdiMinus, mdiAccount, mdiPencil } from '@mdi/js';
import { createRipple } from '../../utils/utils';

/**
 * @group Buttons
 * @slot - Button text
 * @slot icon - Icon
 */
@Component({
  tag: 'c-button',
  styleUrl: 'c-button.scss',
  shadow: true,
})
export class Button {
  /**
   * Outlined button style
   */
  @Prop() outlined = false;

  /**
   * Light button background
   */
  @Prop() ghost = false;

  /**
   * Transparent button background
   */
  @Prop() text = false;

  /**
   * Display loader on the button
   */
  @Prop() loading = false;

  /**
   * Fit width to containing element
   */
  @Prop() fit = false;

  /**
   * Remove the default border radius
   */
  @Prop() noRadius = false;

  /**
   * Disable the button
   */
  @Prop({ reflect: true }) disabled = false;

  /**
   * Name of the icon to be displayed in the button
   *
   * @deprecated Please use the icon slot instead
   */
  @Prop() icon: 'plus' | 'minus' | 'account' | 'edit';

  /**
   * Value for the button
   * - for use in the c-content-switcher
   */
  @Prop() value?: number | string;

  /**
   * Id of the button
   */
  @Prop({ attribute: 'id' }) hostId: string;

  /**
   * Size of the button
   */
  @Prop() size: 'default' | 'small' | 'large' = 'default';

  private _container?: HTMLDivElement;

  private _onClick = (event) => {
    createRipple(event, this._container);
  };

  render() {
    const SPINNER_SMALL: SVGImageElement = (
      <svg
        class="spinner"
        viewBox="0 0 100 100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle class="spinner__circle" cx="50" cy="50" r="45" />
      </svg>
    );

    let selectedIcon = null;
    let svg: SVGImageElement;

    if (this.icon) {
      const icons = {
        plus: mdiPlus,
        minus: mdiMinus,
        account: mdiAccount,
        edit: mdiPencil,
      };
      selectedIcon = icons[this.icon];
      svg = (
        <svg class="button-icon" width="16" height="16" viewBox="0 0 22 22">
          <path d={selectedIcon} />
        </svg>
      );
    }

    const buttonClasses = {
      'c-button': true,
      'c-button--disabled': !!this.disabled,
      'c-button--ghost': !!this.ghost,
      'c-button--small': this.size === 'small',
      'c-button--large': this.size === 'large',
      'c-button--text': !!this.text,
      'c-button--fitted': !!this.fit,
      'c-button--outlined': !!this.outlined,
      'c-button--no-radius': !!this.noRadius,
    };

    const innerClasses = {
      'c-button-padding': true,
      'hide-text': this.loading,
    };

    const hostClasses = {
      fit: !!this.fit,
      'no-radius': !!this.noRadius,
    };

    return (
      <button
        id={this.hostId}
        class={hostClasses}
        tabindex="0"
        role="button"
        disabled={this.disabled}
        onClick={this._onClick}
      >
        <div
          class={buttonClasses}
          ref={(el) => (this._container = el as HTMLDivElement)}
        >
          {this.loading && <div class="spinner_wrapper">{SPINNER_SMALL}</div>}
          <div class={innerClasses}>
            <slot name="icon"></slot>
            {svg}
            <slot></slot>
          </div>
        </div>
      </button>
    );
  }
}
