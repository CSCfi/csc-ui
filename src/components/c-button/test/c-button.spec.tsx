import { newSpecPage } from '@stencil/core/testing';
import { CButton } from '../c-button';

describe('button', () => {
  it('renders default html', async () => {
    const page = await newSpecPage({
      components: [CButton],
      html: `<c-button>Default</c-button>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
