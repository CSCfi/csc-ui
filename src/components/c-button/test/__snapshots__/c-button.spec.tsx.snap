// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button add href 1`] = `
<c-button href="www.google.com">
  <mock:shadow-root>
    <a href="www.google.com" role="button" tabindex="0" target="_blank">
      <div class="c-button">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </a>
  </mock:shadow-root>
</c-button>
`;

exports[`button add id 1`] = `
<c-button id="important">
  <mock:shadow-root>
    <button id="important" role="button" tabindex="0">
      <div class="c-button">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
</c-button>
`;

exports[`button render icon after text 1`] = `
<c-button icon-end="{true}">
  <mock:shadow-root>
    <button role="button" tabindex="0">
      <div class="c-button">
        <div class="c-button__content">
          <slot></slot>
          <slot name="icon"></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
  ButtonText
</c-button>
`;

exports[`button render icon by attribute, not slot 1`] = `
<c-button icon="account">
  <mock:shadow-root>
    <button role="button" tabindex="0">
      <div class="c-button">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <svg class="button-icon" height="16" viewBox="0 0 22 22" width="16">
            <path d="M12,4A4,4 0 0,1 16,8A4,4 0 0,1 12,12A4,4 0 0,1 8,8A4,4 0 0,1 12,4M12,14C16.42,14 20,15.79 20,18V20H4V18C4,15.79 7.58,14 12,14Z"></path>
          </svg>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
</c-button>
`;

exports[`button renders button with filled slots 1`] = `
<c-button>
  <mock:shadow-root>
    <button role="button" tabindex="0">
      <div class="c-button c-button--description">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
        <div class="c-button__description">
          <slot name="description"></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
  <div class="mdi mdi-login" slot="icon"></div>
  Default
  <div slot="description">
    Log out
  </div>
</c-button>
`;

exports[`button renders default html 1`] = `
<c-button>
  <mock:shadow-root>
    <button role="button" tabindex="0">
      <div class="c-button">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
  Default
</c-button>
`;

exports[`button renders disabled button 1`] = `
<c-button disabled="">
  <mock:shadow-root>
    <button disabled="" role="button" tabindex="-1">
      <div class="c-button c-button--disabled">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
</c-button>
`;

exports[`button renders fitted button 1`] = `
<c-button fit="{true}">
  <mock:shadow-root>
    <button class="fit" role="button" tabindex="0">
      <div class="c-button c-button--fitted">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
</c-button>
`;

exports[`button renders ghost button 1`] = `
<c-button ghost="{true}">
  <mock:shadow-root>
    <button role="button" tabindex="0">
      <div class="c-button c-button--ghost">
        <div class="c-button__content">
          <slot name="icon"></slot>
          <slot></slot>
        </div>
      </div>
    </button>
  </mock:shadow-root>
</c-button>
`;
